### ISSUE TITLE
Busybox issue

### ISSUE THREAD

**Issue by timdaman** [bats-core/bats-core#83]
I am running Alpine Linux and when I clone the master branch of bats-core and run `install.sh` I get an error. Here is an example of the output I am seeing.

```
$ cat /etc/alpine-release 
3.7.0

$ ls
CODE_OF_CONDUCT.md  Dockerfile          README.md           install.sh          man                 test
CONTRIBUTING.md     LICENSE.md          bin                 libexec             package.json

$ sudo ./install.sh /usr/local/
readlink: unrecognized option: e
BusyBox v1.27.2 (2018-01-29 15:47:03 UTC) multi-call binary.

Usage: readlink [-fnv] FILE

Display the value of a symlink

        -f      Canonicalize by following all symlinks
        -n      Don't add newline
        -v      Verbose

$ ls -l $(which readlink)
lrwxrwxrwx    1 root     root            12 Feb 17 18:00 /usr/bin/readlink -> /bin/busybox

```

**Comment by timdaman** [comment]
More information on this. To work around the issue add the coreutils package. 
`apk add coreutils`
Still, I think bats should probably identify this issue and present output identifying the cause.

**Comment by bvberkum** [comment]
I tried to reproduce this on a release from early may (9b33c4959ff) but I could not
```
docker run -it --volume $(pwd -P):/opt/bats alpine:3.6 sh -c 'apk --no-cache add bash && cd /opt/bats/ && ./install.sh /usr/local && bats test/bats.bats'
```
There are two calls to readlink, and I may not trigger the second call. It looks like the installer did some symlink fixing on bin/bats. That step is gone for now (but I'm not sure if it stays like that: #90)

**Comment by timdaman** [comment]
Huh, weird. I agree, the issue is not happening in your test. I noticed I was using a newer Alpine in my report but when I tested that made no difference. Sorry to trouble you and thanks for looking into.

**Comment by espoelstra** [comment]
I was bit by this as well when trying to test a script where I was using `readlink -e` after learning about some (IMO terrible and crazy) behavior in `readlink -f` related to non-existent paths not returning the expected error exit code. Since `busybox` provides `readlink` if it isn't in the system I got the same usage warning and had to do some digging into why I was using `-e` in the first place, that opened a whole can of worms but I believe using `-e` is far better than `-f` in most cases.

See this Q&A for some eye opening reading, https://unix.stackexchange.com/a/246987/103652. 
TL;DR The trouble is `busybox` opted to partially combine the `-e` and `-f` options of GNU readlink resulting in the tricky case that if run under busybox a test like `readlink -f non-existent-file` returns `1`, but the same command under GNU readlink on Ubuntu or most other distributions returns `0`.

Snippet from `man readlink` on Ubuntu, note `-f` doesn't actually require the path it is returning to exist

      -f, --canonicalize
           canonicalize by following every symlink in every component of the given name recursively;
           all but the last component must exist
    
      -e, --canonicalize-existing
           canonicalize by following every symlink in every component of the given name recursively,
           all components must exist

My suggestion would be to add `coreutils` into the `bats` Docker container because the behavior of `readlink -f` is not exactly intuitive and using `readlink -e` is probably more correct. Not to mention the `greadlink` mentioned above is from `brew install coreutils` which is one of the packages that OSX Bash hackers probably use along with `brew install findutils` to get their `greadlink` and `gxargs` to have the flags they commonly use available again.

This discrepancy is kind of silly because the Busybox guys combined `-e` and `-f` in their re-implementation of `readlink` because they couldn't figure out how to fully implement `-e` properly. https://github.com/brgl/busybox/blob/master/coreutils/readlink.c#L59

**Comment by sublimino** [comment]
Thanks for the deep dive @espoelstra 

So existing/passing tests using `readlink -f non-existent-file` will fail if run in Docker after adding `coreutils`. This also extends to every other package that `coreutils` installs that differs from the `busybox` installation (I don't know how big that is as a problem):

<details>

```
$ docker run -it --entrypoint="" bats/bats:1  sh -c "apk --update add coreutils && apk info coreutils --contents"
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.6/community/x86_64/APKINDEX.tar.gz
(1/3) Installing libattr (2.4.47-r6)
(2/3) Installing libacl (2.2.52-r3)
(3/3) Installing coreutils (8.27-r0)
Executing busybox-1.26.2-r9.trigger
OK: 18 MiB in 19 packages
coreutils-8.27-r0 contains:
bin/nice
bin/stat
bin/stty
bin/mknod
bin/chgrp
bin/ls
bin/touch
bin/rm
bin/date
bin/base64
bin/mv
bin/sync
bin/cat
bin/mkdir
bin/echo
bin/false
bin/sleep
bin/df
bin/uname
bin/chown
bin/chmod
bin/true
bin/mktemp
bin/cp
bin/pwd
bin/ln
bin/printenv
bin/dd
bin/rmdir
usr/sbin/chroot
usr/libexec/coreutils/libstdbuf.so
usr/bin/tail
usr/bin/expand
usr/bin/sha256sum
usr/bin/users
usr/bin/unlink
usr/bin/tac
usr/bin/shuf
usr/bin/md5sum
usr/bin/tty
usr/bin/pathchk
usr/bin/b2sum
usr/bin/cut
usr/bin/paste
usr/bin/split
usr/bin/base32
usr/bin/basename
usr/bin/env
usr/bin/pr
usr/bin/groups
usr/bin/tr
usr/bin/join
usr/bin/expr
usr/bin/pinky
usr/bin/seq
usr/bin/sha1sum
usr/bin/uniq
usr/bin/dir
usr/bin/dircolors
usr/bin/logname
usr/bin/runcon
usr/bin/link
usr/bin/vdir
usr/bin/whoami
usr/bin/sha512sum
usr/bin/stdbuf
usr/bin/head
usr/bin/comm
usr/bin/od
usr/bin/numfmt
usr/bin/ptx
usr/bin/nl
usr/bin/wc
usr/bin/fold
usr/bin/fmt
usr/bin/who
usr/bin/tee
usr/bin/test
usr/bin/printf
usr/bin/id
usr/bin/unexpand
usr/bin/shred
usr/bin/hostid
usr/bin/realpath
usr/bin/mkfifo
usr/bin/[
usr/bin/sum
usr/bin/du
usr/bin/tsort
usr/bin/readlink
usr/bin/factor
usr/bin/install
usr/bin/dirname
usr/bin/timeout
usr/bin/yes
usr/bin/chcon
usr/bin/sha384sum
usr/bin/nohup
usr/bin/nproc
usr/bin/csplit
usr/bin/sort
usr/bin/truncate
usr/bin/sha224sum
usr/bin/cksum
```
</p>
</details>

We could treat this as a "warn and see who screams" breaking change in the next minor release? /cc @mbland @jasonkarns

**Comment by mbland** [comment]
If I'm understanding correctly: This is a request to add a package to the Docker image not because Bats needs it, but because a significant number of users might?

I'd be hesitant to put it into our Docker image—I like the idea of having Bats run clean on the barest of bare bones Alpine system. The use of `readlink` in bats is confined to `bin/bats` (it was stripped from `install.sh` in #92, which was updated to its current state in #105) and uses `cd` to avoid using `-e` or `-f`. That said, since the busybox implementation is nonstandard, I can see every potential user installing coreutils out of the box for consistency's sake with most every other common Linux distribution.

But the thing is, our image is based on the official Bash image, which already has coreutils installed. So, as far as I can tell, there's nothing further for us to do on this.

**Comment by sublimino** [comment]
> our image is based on the official Bash image, which already has coreutils installed

I can't verify this, either with the code dump in the details tab of https://github.com/bats-core/bats-core/issues/83#issuecomment-425024432 or with the latest `bash` image:

```
$ docker run -it bash readlink --help
BusyBox v1.28.4 (2018-07-17 15:21:40 UTC) multi-call binary.
```

vs

```
$ docker run -it bash -c "apk --update add coreutils && readlink --help"                                                                                      

fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.8/community/x86_64/APKINDEX.tar.gz
(1/3) Installing libattr (2.4.47-r7)
(2/3) Installing libacl (2.2.52-r5)
(3/3) Installing coreutils (8.29-r2)
Executing busybox-1.28.4-r1.trigger
OK: 13 MiB in 20 packages
Usage: readlink [OPTION]... FILE...
Print value of a symbolic link or canonical file name
<snip>
GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
```

> I can see every potential user installing coreutils out of the box for consistency's sake with most every other common Linux distribution.

I'm inclined to agree.
